vector/src/main/java/im/vector/app/features/widgets/webview/WebviewPermissionUtils.kt

```diff
@@ -37,16 +50,50 @@ object WebviewPermissionUtils {
                     allowedPermissions[which] = allowedPermissions[which].first to isChecked
                 }
                 .setPositiveButton(R.string.room_widget_resource_grant_permission) { _, _ ->
-                    request.grant(allowedPermissions.mapNotNull { perm ->
+                    permissionRequest = request
+                    selectedPermissions = allowedPermissions.mapNotNull { perm ->
                         perm.first.takeIf { perm.second }
-                    }.toTypedArray())
+                    }
+
+                    val requiredAndroidPermissions = selectedPermissions.mapNotNull { permission ->
+                        webPermissionToAndroidPermission(permission)
+                    }
+
+                    // When checkPermissions returns false, some of the required Android permissions will
+                    // have to be requested and the flow completes asynchronously via onPermissionResult
+                    if (checkPermissions(requiredAndroidPermissions, activity, activityResultLauncher)) {
+                        request.grant(selectedPermissions.toTypedArray())
+                        reset()
+                    }
                 }
                 .setNegativeButton(R.string.room_widget_resource_decline_permission) { _, _ ->
                     request.deny()
                 }
                 .show()
     }
 
+    fun onPermissionResult(result: Map<String, Boolean>) {
+        permissionRequest?.let { request ->
+            val grantedPermissions = selectedPermissions.filter { webPermission ->
+                val androidPermission = webPermissionToAndroidPermission(webPermission)
+                        ?: return@filter true // No corresponding Android permission exists
+                return@filter result[androidPermission]
+                        ?: return@filter true // Android permission already granted before
+            }
+            if (grantedPermissions.isNotEmpty()) {
```

Comment:
Hm, I think in this snippet we'd put the _Android_ permission strings into `request.grant(...)`, right? We have to put the _web_ permission strings into that one. That's essentially the reason why I mapped inside the filter.

Reply:
I agree that the `filter` clause isn't very easy to parse though. ☹️

Reply:
This kept nagging me so I added a unit test for this part to be able to safely change the implementation. I found that your updated snippet still doesn't work, sadly.

First, `onlySupportedAndroidPermissions` is actually not needed / desired because there are web permissions such as `PermissionRequest.RESOURCE_PROTECTED_MEDIA_ID` for which (I think) no Android system permission exists. In those cases we still want to grant those permissions to the web view (based on the fact that the user confirmed them in the _widget_ permission dialog) so we can't filter them out.

Secondly, the chained call of `mapWithResult` uses the web permission string as key in the _Android_ permission result map which will always result in `null`.

I've tried to resolve this but in the end it comes down to having the current complicated filter/map logic inside `mapWithResult` which seemed more or less equivalent to what I have now. 😕

At least we have the tests now though so this endeavor hasn't been in vein. 😀